dask/array/core.py

```diff
@@ -1007,6 +1003,41 @@ def __new__(cls, dask, name, chunks, dtype, shape=None):
     def __reduce__(self):
         return (Array, (self.dask, self.name, self.chunks, self.dtype))
 
+    def __dask_graph__(self):
+        return self.dask
+
+    def __dask_keys__(self):
+        if self._cached_keys is not None:
+            return self._cached_keys
+
+        name, chunks, numblocks = self.name, self.chunks, self.numblocks
+
+        def keys(*args):
+            if not chunks:
+                return [(name,)]
+            ind = len(args)
+            if ind + 1 == len(numblocks):
+                result = [(name,) + args + (i,) for i in range(numblocks[ind])]
+            else:
+                result = [keys(*(args + (i,))) for i in range(numblocks[ind])]
+            return result
```

Comment:
Why the nested function here?

Reply:
We may want a deprecation cycle around `_keys`.  I would not be surprised if this has leaked out into semi-public API.

Reply:
Yeah have even mucked with `_keys` some myself to optimize things. ( https://github.com/dask/dask/issues/2472 ) That said, the general rule of thumb is that `_` means not public.

Reply:
> Why the nested function here

Array keys are implemented recursively. Since `__dask_keys__` takes no parameters I moved the recursive function to be internal. Could also be moved external to a helper function, but this seemed cleaner and keeps things located together.

> We may want a deprecation cycle around `_keys`:

All the old methods still exist and work, and will issue a deprecation warning on use. `_keys` calls `__dask_keys__`.